### PR TITLE
[SR-538][Stdlib][3.0] Updated constraints and docs for popFirst/popLast. Added tests.

### PR DESCRIPTION
**Pull request by tanadeau**
3.0 piece of SR-538. When/if this is accepted, I can backport the changes to 2.2.

### CODE REVIEW

File: stdlib/public/core/Collection.swift [hunk 1]
```diff
@@ -234,17 +234,19 @@ extension CollectionType where SubSequence == Self {
     self = self[startIndex.successor()..<endIndex]
     return element
   }
+}
 
+extension CollectionType where
+    SubSequence == Self, Index: BidirectionalIndexType {
```
Comment (gribozavr) [inline stdlib/public/core/Collection.swift]: Please add a space after the colon.

Comment (tanadeau) [inline stdlib/public/core/Collection.swift]: Will do. Is there a doc for stdlib coding conventions (indentation, etc.)? I couldn't find one.

**Comment by gribozavr** [comment]
Thank you!

Could you add the tests to `stdlib/private/StdlibUnittest/CheckCollection.swift` instead?  This way, they will get run on all collections, not just arrays, and it will check that the implementation does indeed work on a strict collection implementation that has minimal guarantees.

`popFirst()` tests will go into `addForwardCollectionTests`.  You can model them after `removeFirst()` tests (and even reuse the same data set + add a test for an empty collection).

`popLast()` tests will go into `addBidirectionalCollectionTests()`.  These would be similar to `removeLast()` tests (again, you can reuse the same dataset).

File: stdlib/public/core/Collection.swift (hunk 1 above)
Comment (gribozavr) [inline stdlib/public/core/Collection.swift]: I'm afraid we don't have it written down.  Please imitate the coding style of the code you see in the library.

**Comment by tanadeau** [comment]
I moved the tests as suggested and rebased against latest master.

**Comment by gribozavr** [comment]
Thank you!  Tested on OS X, tests pass.  Merging.

**Comment by gribozavr** [comment]
> When/if this is accepted, I can backport the changes to 2.2.

This would be awesome, thank you!

**Comment by tanadeau** [comment]
Backport in #1086

**Pull request merged**